In [1]:
from glob import glob
import os, sys
from DNASkittleUtils.Contigs import *

In [203]:
from natsort import natsorted
a = ['2 ft 7 in', '1 ft 5 in', '10 ft 2 in', '2 ft 11 in', '7 ft 6 in']
natsorted(a)

['1 ft 5 in', '2 ft 7 in', '2 ft 11 in', '7 ft 6 in', '10 ft 2 in']

In [204]:
def csv_file_into_data_table(count_path):
    with open(count_path, 'r') as count_file:
        r = csv.reader(count_file, delimiter=':')
        frax_hogs = [line for line in r]
        counts = [int(x[1]) for x in frax_hogs]
        names = [x[0].replace('.fa', '') for x in frax_hogs]
    return counts, names

def frax_gene_counts_dataframe(directory):
    import pandas as pd
    import os
    file_list = glob(os.path.join(directory, "*"))
    assert len(file_list), f"Directory '{directory}' is empty or doesn't exist"
    
    all_counts = []
    ignored, names = csv_file_into_data_table(file_list[0])
    for filename in file_list:
        counts, current_names = csv_file_into_data_table(filename)
        assert current_names == names
        all_counts.append(counts)
    
    species_labels = [just_the_name(f).replace('number_of_', '').replace('_sequences_in_each_HOG', '') for f in file_list]
    
    d = {}
    assert len(species_labels) == len(all_counts)
    for i in range(len(species_labels)):
        d[species_labels[i]] = pd.Series(all_counts[i], index=names) 
    data_table = pd.DataFrame(d)
    data_table = data_table.reindex(natsorted(names))
    return data_table

In [211]:
count_path = r"D:\josiah\Documents\Research\Thesis - Genome Symmetry\DNA_Duplications\data\per species HOG count"
df = frax_gene_counts_dataframe(count_path)

In [212]:
df['HOG10':'HOG100']

FEXC  FRAX01  FRAX03  FRAX04  FRAX05  FRAX06  FRAX07  FRAX08  FRAX09  \
HOG10      1       1       1       1       1       1       1       1       1   
HOG11      2       2       2       3       3       2       2       4       2   
HOG12      1       1       1       0       0       1       1       0       1   
HOG13      1       1       1       3       0       1       1       2       1   
HOG14      1       1       1       1       2       1       1       1       1   
HOG15      1       1       1       1       1       1       1       1       1   
HOG16      3       3       3       3       3       3       3       3       3   
HOG17      0       1       0       0       0       1       1       1       0   
HOG18      1       1       1       1       1       1       1       1       1   
HOG19      2       2       2       2       2       2       2       2       2   
HOG20      1       1       1       2       1       1       1       2       1   
HOG21      2       2       2       1       2       2       2       1       0   
HOG22      2       2       2       2       2       2       2       2       2   
HOG23      1       1       0       1       1       1       1       1       1   
HOG24      2       2       3       2       2       2       2       2       2   
HOG25      3       3       3       3       3       3       3       3       3   
HOG26      3       3       4       4       4       3       3       3       3   
HOG27      1       1       1       1       1       1       1       1       1   
HOG28     20      20      20      20      20      20      20      20      20   
HOG29      1       1       1       1       1       1       1       1       1   
HOG30     11      11      12      10      10      11      12      12      12   
HOG31      3       3       3       3       3       3       3       3       3   
HOG32      1       1       1       1       1       1       1       1       1   
HOG33      1       1       1       1       1       1       1       1       1   
HOG34      1       1       1       1       1       1       1       1       1   
HOG35      1       1       1       1       1       1       1       1       1   
HOG36      1       1       1       1       1       1       1       1       1   
HOG37      1       1       1       1       1       1       1       1       1   
HOG38      3       3       3       3       3       3       3       2       3   
HOG39      0       1       1       1       1       1       1       1       1   
...      ...     ...     ...     ...     ...     ...     ...     ...     ...   
HOG71      2       2       2       2       2       2       2       2       2   
HOG72      3       3       3       3       3       3       3       3       3   
HOG73      0       0       0       0       0       0       1       0       0   
HOG74      2       2       2       2       1       2       2       2       2   
HOG75      1       1       1       1       2       1       1       1       1   
HOG76      1       1       1       1       1       1       1       1       1   
HOG77      4       4       4       4       4       3       4       4       4   
HOG78      0       1       1       1       1       0       0       1       0   
HOG79      0       0       0       0       0       1       0       0       0   
HOG80      1       2       0       1       0       0       1       1       0   
HOG81      2       2       1       2       2       2       2       2       2   
HOG82      1       1       2       1       1       1       1       1       1   
HOG83      1       1       1       1       1       1       1       1       1   
HOG84      1       1       1       1       1       1       1       1       1   
HOG85      3       5       3       5       5       4       4       5       3   
HOG86      3       0       2       1       0       2       1       0       2   
HOG87      5       5       5       5       5       5       5       5       5   
HOG88      2       2       1       1       2       2       2       2       2   
HOG89   

### Write File!

In [213]:
df.to_csv('data/HOG_counts_table.csv', sep='\t')

PermissionError: [Errno 13] Permission denied: 'data/HOG_counts_table.csv'

# Exploration

### Questions we'd like to answer:

* How many HOGs have 32 species?
* How many have all Fraxinus species?
* What is the largest group?
* How many HOGs are going to have a usable signal for duplication and loss

In [218]:
[row for row in df.itertuples()]

[Pandas(Index='HOG1', FEXC=2, FRAX01=2, FRAX03=2, FRAX04=2, FRAX05=2, FRAX06=2, FRAX07=2, FRAX08=2, FRAX09=2, FRAX10=2, FRAX11=2, FRAX12=2, FRAX13=2, FRAX14=2, FRAX15=2, FRAX16=2, FRAX19=2, FRAX20=2, FRAX21=2, FRAX23=2, FRAX25=2, FRAX26=2, FRAX27=2, FRAX28=2, FRAX29=2, FRAX30=2, FRAX31=2, FRAX32=2, FRAX33=2, Mimulus=1, Olea=3, Solanum=1),
 Pandas(Index='HOG2', FEXC=2, FRAX01=2, FRAX03=2, FRAX04=2, FRAX05=2, FRAX06=2, FRAX07=2, FRAX08=2, FRAX09=2, FRAX10=2, FRAX11=2, FRAX12=2, FRAX13=2, FRAX14=2, FRAX15=2, FRAX16=2, FRAX19=2, FRAX20=2, FRAX21=2, FRAX23=2, FRAX25=2, FRAX26=2, FRAX27=2, FRAX28=2, FRAX29=2, FRAX30=2, FRAX31=2, FRAX32=2, FRAX33=2, Mimulus=0, Olea=3, Solanum=0),
 Pandas(Index='HOG3', FEXC=1, FRAX01=1, FRAX03=1, FRAX04=1, FRAX05=1, FRAX06=1, FRAX07=1, FRAX08=1, FRAX09=1, FRAX10=1, FRAX11=1, FRAX12=1, FRAX13=1, FRAX14=0, FRAX15=1, FRAX16=1, FRAX19=1, FRAX20=1, FRAX21=1, FRAX23=1, FRAX25=1, FRAX26=1, FRAX27=1, FRAX28=1, FRAX29=1, FRAX30=1, FRAX31=1, FRAX32=1, FRAX33=1, Mimulus=

In [225]:
na = df.values
na.sum()

1181232

In [237]:
row_totals = df.sum(axis=1)
len(row_totals)

28794

### What is the largest group?

In [303]:
max(row_totals)

1089

### Highest Gene copy number in 1 genome

In [240]:
df.max()

FEXC       35
FRAX01     35
FRAX03     35
FRAX04     35
FRAX05     35
FRAX06     35
FRAX07     35
FRAX08     35
FRAX09     35
FRAX10     35
FRAX11     35
FRAX12     34
FRAX13     35
FRAX14     34
FRAX15     34
FRAX16     35
FRAX19     35
FRAX20     35
FRAX21     35
FRAX23     35
FRAX25     35
FRAX26     35
FRAX27     34
FRAX28     35
FRAX29     35
FRAX30     35
FRAX31     35
FRAX32     35
FRAX33     35
Mimulus    86
Olea       41
Solanum    31
dtype: int64

35 is suspiciously close to 32.  This is also the largest group we're likely picking up here because 35 * 32 ~= 1089

## At least 1 per species of FRAX (only)

In [255]:
import numpy as np

In [266]:
df.values[25000], np.count_nonzero(df.values[25000])

(array([0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        1, 1, 0, 0, 0, 0, 0, 0, 0], dtype=int64), 6)

In [257]:
at_least_one = [np.count_nonzero( row) for row in df.values]

In [263]:
from statistics import mean
mean(at_least_one)

25.207091755226784

In [268]:
col_list = list(df)
col_list.remove('Mimulus')
col_list.remove('Olea')
col_list.remove('Solanum')
just_frax = df[col_list]

In [269]:
at_least_one = [np.count_nonzero( row) for row in just_frax.values]

In [270]:
from statistics import mean
mean(at_least_one)

23.486976453427797

In [272]:
max(at_least_one)

29

In [301]:
from collections import Counter
cov_counts = Counter(at_least_one)

In [286]:
15130 / sum(cov_counts.values())

0.5254566923664652

In [289]:
['{:.2%}'.format( cov_counts[i]  / sum(cov_counts.values())) for i in range(30)]

['2.93%',
 '0.90%',
 '4.96%',
 '2.14%',
 '1.43%',
 '0.99%',
 '0.85%',
 '0.67%',
 '0.51%',
 '0.44%',
 '0.44%',
 '0.42%',
 '0.41%',
 '0.38%',
 '0.41%',
 '0.36%',
 '0.38%',
 '0.49%',
 '0.48%',
 '0.57%',
 '0.70%',
 '0.87%',
 '1.06%',
 '1.25%',
 '1.65%',
 '2.12%',
 '3.14%',
 '5.41%',
 '11.09%',
 '52.55%']

In [276]:
from bokeh.io import output_notebook, show
output_notebook()

Loading BokehJS ...

In [293]:
cov_counts.keys(), cov_counts.values()

(dict_keys([29, 28, 21, 26, 23, 13, 27, 24, 12, 25, 1, 0, 7, 6, 2, 16, 3, 19, 22, 20, 8, 11, 4, 10, 5, 14, 15, 17, 18, 9]),
 dict_values([15130, 3193, 250, 905, 359, 109, 1559, 475, 117, 609, 260, 844, 194, 244, 1427, 109, 616, 165, 306, 202, 147, 121, 411, 127, 286, 119, 105, 140, 138, 127]))

dict.keys() is not reliably ordered

In [300]:
from bokeh.plotting import figure
f = figure(title='Species coverage of HOGs Histogram')
x = list(range(30))
f.line(x, [cov_counts[i] for i in x])
show(f   )

# Acceptable HOGs
* Present in at least half of the species
* Std Dev higher than 2
* max greater than 3

In [341]:
species_coverage = pd.Series([np.count_nonzero( row) for row in just_frax.values], index=just_frax.index)

In [304]:
row_totals = just_frax.sum(axis=1)

In [305]:
mean(row_totals)

38.215218448287835

In [337]:
mean(at_least_one)

23.486976453427797

Average HOG has 38 genes in 23 FRAX species

**Per row standard deviation**

In [312]:
variation = just_frax.std(axis=1)
variation

HOG1        0.000000
HOG2        0.000000
HOG3        0.185695
HOG4        0.000000
HOG5        0.371391
HOG6        0.185695
HOG7        0.454859
HOG8        0.185695
HOG9        0.619868
HOG10       0.000000
HOG11       0.510964
HOG12       0.412251
HOG13       0.751026
HOG14       0.488791
HOG15       0.185695
HOG16       0.000000
HOG17       0.506120
HOG18       0.257881
HOG19       0.000000
HOG20       0.384426
HOG21       0.613879
HOG22       0.000000
HOG23       0.658405
HOG24       0.658405
HOG25       0.000000
HOG26       0.501230
HOG27       0.000000
HOG28       0.000000
HOG29       0.000000
HOG30       0.737791
              ...   
HOG28765    0.000000
HOG28766    0.000000
HOG28767    0.000000
HOG28768    0.000000
HOG28769    0.000000
HOG28770    0.000000
HOG28771    0.000000
HOG28772    0.000000
HOG28773    0.000000
HOG28774    0.000000
HOG28775    0.000000
HOG28776    0.000000
HOG28777    0.000000
HOG28778    0.000000
HOG28779    0.000000
HOG28780    0.000000
HOG28781    0

In [311]:
just_frax.loc['HOG1']

FEXC      2
FRAX01    2
FRAX03    2
FRAX04    2
FRAX05    2
FRAX06    2
FRAX07    2
FRAX08    2
FRAX09    2
FRAX10    2
FRAX11    2
FRAX12    2
FRAX13    2
FRAX14    2
FRAX15    2
FRAX16    2
FRAX19    2
FRAX20    2
FRAX21    2
FRAX23    2
FRAX25    2
FRAX26    2
FRAX27    2
FRAX28    2
FRAX29    2
FRAX30    2
FRAX31    2
FRAX32    2
FRAX33    2
Name: HOG1, dtype: int64

HOG1 has 0 std dev because they're all 2

In [376]:
enough_variation = variation > .5

len(enough_variation.nonzero()[0])

3476

In [386]:
at_least_fourteen_species = species_coverage > (29//2)

acceptable_values = enough_variation & at_least_fourteen_species

len(acceptable_values.nonzero()[0])

2961

In [387]:
filtered_candidates = just_frax.iloc[acceptable_values.nonzero()[0]]
filtered_candidates

FEXC  FRAX01  FRAX03  FRAX04  FRAX05  FRAX06  FRAX07  FRAX08  \
HOG9         2       2       0       2       2       2       2       0   
HOG11        2       2       2       3       3       2       2       4   
HOG13        1       1       1       3       0       1       1       2   
HOG21        2       2       2       1       2       2       2       1   
HOG23        1       1       0       1       1       1       1       1   
HOG24        2       2       3       2       2       2       2       2   
HOG26        3       3       4       4       4       3       3       3   
HOG30       11      11      12      10      10      11      12      12   
HOG43        2       2       1       1       1       2       1       1   
HOG44        4       4       4       4       4       4       4       4   
HOG52        4       5       7       3       6       4       4       4   
HOG53        1       0       1       1       0       1       2       2   
HOG58        1       1       2       1       1       1       1       1   
HOG67        2       1       1       2       0       2       2       2   
HOG74        2       2       2       2       1       2       2       2   
HOG75        1       1       1       1       2       1       1       1   
HOG78        0       1       1       1       1       0       0       1   
HOG80        1       2       0       1       0       0       1       1   
HOG85        3       5       3       5       5       4       4       5   
HOG86        3       0       2       1       0       2       1       0   
HOG94        1       0       1       1       1       1       1       3   
HOG122       1       1       0       0       0       1       2       0   
HOG127       2       2       1       3       3       2       2       2   
HOG133       2       2       1       3       3       3       1       1   
HOG134       2       3       2       2       2       3       3       3   
HOG139       2       3       4       2       4       2       2       3   
HOG140       2       0       0       2       0       2       2       2   
HOG161       3       3       3       3       0       3       3       3   
HOG165       2       4       2       3       4       3       3       3   
HOG172       6       6       6       6       6       5       6       6   
...        ...     ...     ...     ...     ...     ...     ...     ...   
HOG26077     1       0       0       0       1       0       1       1   
HOG26084     1       0       0       0       2       1       1       1   
HOG26106     1       0       0       0       1       1       1       1   
HOG26113     1       0       0       0       2       0       1       1   
HOG26119     1       0       0       0       1       1       0       0   
HOG26125     1       0       0       0       1       1       1       1   
HOG26150     1       0       0       0       1       1       1       1   
HOG26168     1       0       0       0       1       1       1       0   
HOG26231     1       0       0       0       1       0       0       0   
HOG26243     1       0       0       0       0       1       1       1   
HOG26249     1       0       0       0       0       1       0       0   
HOG26271     3       0       0       0       0       3       3       0   
HOG26273     1       0       0       0       0       1       0       0   
HOG26276     2       0       0       0       0       2       2       0   
HOG26284     1       0       0       0       0       1       1       1   
HOG26322     1       0       0       0       0       1       1       0   
HOG26323     1       0       0       0       0       1       1       1   
HOG26335     1       0       0       0       0       1       1       0   
HOG26343     1       0       0       0       0       1       1       0   
HOG26356     1       0       0       0       0       1       1       1   
HOG26358     1       0       0       0       0       1       1       0   
HOG26359     1       0       0       0       0       1       1       1   
HOG26362     1

In [388]:
filtered_candidates.to_csv('data/candidate_HOGs_for_study.csv', sep='\t')

In [391]:
with open('copy_candidate_HOGs.sh', 'w') as out:
    folder = '/data/SBCS-BuggsLab/Josiah/OrthologousGroups/candidates/'
    out.write(f'mkdir {folder}\n')
    for hog in filtered_candidates.index:
        out.write(f'cp {hog}.fa {folder} \n')

### Largest Single HOG = 1089 genes

In [273]:
max(df.sum(axis=1))

1089

In [231]:
col_totals = df.sum(axis=0)
col_totals

FEXC       38248
FRAX01     37955
FRAX03     37773
FRAX04     37743
FRAX05     37749
FRAX06     38165
FRAX07     38113
FRAX08     37633
FRAX09     38094
FRAX10     37699
FRAX11     38101
FRAX12     37564
FRAX13     37838
FRAX14     37992
FRAX15     37967
FRAX16     37930
FRAX19     38065
FRAX20     37775
FRAX21     37809
FRAX23     38046
FRAX25     37737
FRAX26     38154
FRAX27     38093
FRAX28     38035
FRAX29     37952
FRAX30     38089
FRAX31     37970
FRAX32     38041
FRAX33     38039
Mimulus    22065
Olea       37652
Solanum    21146
dtype: int64

In [236]:
 pd.Series(['{:.2%}'.format(  x/ 38248) for x in col_totals], index=col_totals.index)

FEXC       100.00%
FRAX01      99.23%
FRAX03      98.76%
FRAX04      98.68%
FRAX05      98.70%
FRAX06      99.78%
FRAX07      99.65%
FRAX08      98.39%
FRAX09      99.60%
FRAX10      98.56%
FRAX11      99.62%
FRAX12      98.21%
FRAX13      98.93%
FRAX14      99.33%
FRAX15      99.27%
FRAX16      99.17%
FRAX19      99.52%
FRAX20      98.76%
FRAX21      98.85%
FRAX23      99.47%
FRAX25      98.66%
FRAX26      99.75%
FRAX27      99.59%
FRAX28      99.44%
FRAX29      99.23%
FRAX30      99.58%
FRAX31      99.27%
FRAX32      99.46%
FRAX33      99.45%
Mimulus     57.69%
Olea        98.44%
Solanum     55.29%
dtype: object

# Dictionary demo

In [383]:
answer = 2
if answer == 1:
    print("asjklajsd")
elif answer == 2:
    print("adajksldjakl")
elif answer == 2:
    print("ajskd;la" )

adajksldjakl


* Condense any reused into a Function
* use dictionaries instead of if/else

In [12]:
''.join([hex(245)[2:], hex(13)[2:], hex(155)[2:]])

'f5d9b'